File: backend/routers/session.py
from fastapi import APIRouter, HTTPException, Depends
from sqlalchemy.orm import Session
from backend.database.db import get_db
from backend.services.memory import memory_service

router = APIRouter(prefix="/api", tags=["sessions"])

@router.get("/sessions")
def list_sessions_endpoint(db: Session = Depends(get_db)):
    """
    Returns list of all active/past chat sessions with associated PDF details.
    """
    sessions = memory_service.list_user_sessions(db)
    return {
        "status": "success",
        "data": sessions
    }

@router.get("/sessions/{session_id}/messages")
def get_session_messages_endpoint(session_id: str, db: Session = Depends(get_db)):
    """
    Returns full message transcript for a given chat session.
    """
    messages = memory_service.get_session_messages(db, session_id)
    return {
        "status": "success",
        "session_id": session_id,
        "data": messages
    }

@router.delete("/sessions/{session_id}")
def delete_session_endpoint(session_id: str, db: Session = Depends(get_db)):
    """
    Deletes a specified chat session and its saved message transcript.
    """
    success = memory_service.delete_session(db, session_id)
    if not success:
        raise HTTPException(status_code=404, detail="Session not found")
    return {
        "status": "success",
        "message": f"Session {session_id} deleted successfully."
    }

File: backend/src/routers/session.ts
import { Router, Request, Response } from 'express';
import { memoryService } from '../services/memory';

export const sessionRouter = Router();

sessionRouter.get('/sessions', async (req: Request, res: Response): Promise<void> => {
  try {
    const sessions = await memoryService.listUserSessions();
    res.status(200).json({
      status: 'success',
      data: sessions,
    });
  } catch (err: any) {
    res.status(500).json({ detail: err.message || 'Failed to list sessions.' });
  }
});

sessionRouter.get(
  '/sessions/:session_id/messages',
  async (req: Request, res: Response): Promise<void> => {
    try {
      const sessionId = req.params.session_id;
      const messages = await memoryService.getSessionMessages(sessionId);
      res.status(200).json({
        status: 'success',
        session_id: sessionId,
        data: messages,
      });
    } catch (err: any) {
      res.status(500).json({ detail: err.message || 'Failed to get messages.' });
    }
  }
);

sessionRouter.delete(
  '/sessions/:session_id',
  async (req: Request, res: Response): Promise<void> => {
    try {
      const sessionId = req.params.session_id;
      const success = await memoryService.deleteSession(sessionId);
      if (!success) {
        res.status(404).json({ detail: 'Session not found' });
        return;
      }
      res.status(200).json({
        status: 'success',
        message: `Session ${sessionId} deleted successfully.`,
      });
    } catch (err: any) {
      res.status(500).json({ detail: err.message || 'Failed to delete session.' });
    }
  }
);
